### PR TITLE
refactor: drop most CoeFun instances

### PR DESCRIPTION
**Pull request by urkud**
If we define a `CoeFun` instance instead of a `FunLike` instance, then Lean unfolds the definition. We don't want this, at least until we find a way to make it work with `*HomClass`es or find an equally convenient machinery.

- [x] depends on: #1488

### CODE REVIEW

**Comment by kbuzzard** [comment]
My guess is that stuff like this will make `simp` work more like it works in Lean 3. Obviously we don't want to start refactoring mathlib4 before we've ported mathlib3 but these changes look pretty innocuous to me.

**Comment by digama0** [comment]
Ping @gebner . At first glance this seems incorrect - we just need to make sure that any CoeFun instances are wrapped in functions with the `@[coe]` attribute. (PS: I'm on holiday this week and I'm probably not going to be spending much time on lean.)

**Comment by urkud** [comment]
@digama0 Lean 4 does not unfold `FunLike.coe`, so `((f : M →* N) : M → N)` unfolds to `FunLike.coe f`, not to `f.toFun`. This allows us to use generic lemmas like `map_add`, `map_sum` etc.
In `master`, `((o : OrderIso α β) : α → β)`  unfolds to something ugly (AFAIR, it's `FunLike.coe o.toOrderEmbedding.toEmbedding` but I'm not sure). Of course, we can drop `FunLike.coe` and unfold it to `o.toFun` but this will be `o.toEquiv.toFun`, not just `OrderIso.toFun o`. This means that we'll need to replicate, e.g., lemmas about `OrderEmbedding`.
Similarly, without `FunLike`, `((f : M →* N) : M → N)` is `f.toOneHom.toFun` and there is no clear way to apply `MulHom.map_mul` here.

File: Mathlib/Algebra/Hom/Group.lean [hunk 1]
```diff
@@ -598,29 +591,32 @@ initialize_simps_projections MulHom (toFun → apply)
...
 theorem MonoidHom.coe_mk [MulOneClass M] [MulOneClass N] (f hmul) :
   (MonoidHom.mk f hmul : M → N) = f := rfl
 #align monoid_hom.coe_mk MonoidHom.coe_mk
 #align add_monoid_hom.coe_mk AddMonoidHom.coe_mk
 
-@[simp]
-theorem MonoidWithZeroHom.coe_mk [MulZeroOneClass M] [MulZeroOneClass N] (f h1 hmul) :
-  (MonoidWithZeroHom.mk f h1 hmul : M → N) = (f : M → N) := rfl
-#align monoid_with_zero_hom.coe_mk MonoidWithZeroHom.coe_mk
-
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Group.lean:623]: Is there a reason this lemma moved? Does this order make it more or less similar to lean3?

File: Mathlib/Order/Hom/Basic.lean [hunk 2]
```diff
@@ -232,26 +232,16 @@ protected theorem mono (f : α →o β) : Monotone f :=
...
 #align order_hom.mono OrderHom.mono
 
-instance : OrderHomClass (α →o β) α β where
-  coe := toFun
-  coe_injective' f g h := by
-    cases f
-    cases g
-    congr
-  map_rel f _ _ h := f.monotone h
-
```
Comment (eric-wieser) [inline Mathlib/Order/Hom/Basic.lean:242]: This looks like a mistake to me

Comment (eric-wieser) [reply Mathlib/Order/Hom/Basic.lean:242]: Oh nevermind, the proof is above

File: Mathlib/Order/RelIso/Basic.lean [hunk 3]
```diff
@@ -234,19 +227,17 @@ def toRelHom (f : r ↪r s) : r →r s where
 instance : Coe (r ↪r s) (r →r s) :=
   ⟨toRelHom⟩
 
--- see Note [function coercion]
-instance : CoeFun (r ↪r s) fun _ => α → β :=
-  ⟨fun o => o.toEmbedding⟩
-
```
Comment (eric-wieser) [inline Mathlib/Order/RelIso/Basic.lean:240]: Should this have a `#noalign` or porting comment?

**Comment by gebner** [comment]
> At first glance this seems incorrect - we just need to make sure that any CoeFun instances are wrapped in functions with the @[coe] attribute. (PS: I'm on holiday this week and I'm probably not going to be spending much time on lean.)

I think this PR is the right choice.  If we want to have `FunLike` (and now is probably not the right time to explore other designs) then `FunLike.coe` should be the unique function symbol we use for hom-to-fun coercion.

**Comment by digama0** [comment]
Okay, I retract my concern then.

**Comment by urkud** [comment]
I'll merge `master` and go over minor concerns tomorrow.

**Comment by urkud** [comment]
For some reason, the coercion of a `MulEquiv` gets unfolded. I don't know why.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1488~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by urkud** [comment]
I merged `master`. Probably, most of the `diff` is not relevant anymore.

**Comment by urkud** [comment]
All the good parts were moved to other PRs and merged, so I'm closing this one.

**Pull request closed without merging**